styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -60,132 +59,203 @@
     this.storage = Objects.requireNonNull(storage);
     this.eventConsumer = Objects.requireNonNull(eventConsumer);
     this.eventConsumerExecutor = Objects.requireNonNull(eventConsumerExecutor);
-    this.dispatcherThread = new Thread(this::dispatch);
-    dispatcherThread.setName(DISPATCHER_THREAD_NAME);
-    dispatcherThread.start();
+    this.eventTransitionExecutor = Objects.requireNonNull(eventTransitionExecutor);
+    this.outputHandler = Objects.requireNonNull(outputHandler);
   }
 
   @Override
-  public void initialize(RunState runState) throws IsClosedException {
+  public CompletionStage<Void> trigger(WorkflowInstance workflowInstance, Trigger trigger)
+      throws IsClosedException {
     ensureRunning();
-
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new AlreadyInitializedException("RunState initialization called on active instance "
-                                            + workflowInstance.toKey());
-    }
-
     final long counter;
     try {
       counter = storage.getLatestStoredCounter(workflowInstance).orElse(NO_EVENTS_PROCESSED);
     } catch (IOException e) {
-      throw Throwables.propagate(e);
+      return CompletableFutures.exceptionallyCompletedFuture(e);
     }
 
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, counter + 1));
-  }
+    // TODO: optional retry on transaction conflict
 
-  @Override
-  public void restore(RunState runState, long count) {
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new RuntimeException("RunState initialization called on active instance "
-                                 + workflowInstance.toKey());
-    }
+    return CompletableFuture.supplyAsync(() -> {
 
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, count + 1));
+      // Write active state to datastore
+      final RunState runState = RunState.create(workflowInstance, State.NEW, time.get());
+      try {
+        storage.runInTransaction(tx -> {
+          final Optional<Workflow> workflow = tx.workflow(workflowInstance.workflowId());
+          if (!workflow.isPresent()) {
+            throw new IllegalArgumentException("Workflow not found: " + workflowInstance.workflowId().toKey());
+          }
+          return tx.insertActiveState(workflowInstance, PersistentWorkflowInstanceState.of(runState, counter));
+        });
+      } catch (TransactionException e) {
+        if (e.isAlreadyExists()) {
+          throw new IllegalStateException("Workflow instance is already triggered: " + workflowInstance.toKey());
+        } else if (e.isConflict()) {
+          LOG.debug("Transactional conflict, abort triggering Workflow instance: " + workflowInstance.toKey());
+          throw new RuntimeException(e);
```

Comment:
A bit worried what this will look like to the API caller. INTERNAL_SERVER_ERROR, that seems not really right. Something like CONFLICT would be kind of informative as to the root cause, as well as lead users to retrying instead of poking DataEx with bug reports.